### PR TITLE
Add Select Middleware

### PR DESCRIPTION
**Pull request by blackheaven**
Before submitting your PR, check that you've:

- [x] Bumped the version number
- [x] Documented new APIs with [Haddock markup](https://www.haskell.org/haddock/doc/html/index.html)
- [x] Added [`@since` declarations](http://haskell-haddock.readthedocs.io/en/latest/markup.html#since) to the Haddock

After submitting your PR:

- [x] Update the Changelog.md file with a link to your PR
- [x] Check that CI passes (or if it fails, for reasons unrelated to your change, like CI timeouts)

Sorry for the two bumps in two days.

See you tomorrow.

PS: Just kidding

### CODE REVIEW

File: wai-extra/Network/Wai/Middleware/Select.hs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+newtype MiddlewareSelection = MiddlewareSelection
+  { applySelectedMiddleware :: Request -> Maybe Middleware
+  }
+
+instance Semigroup MiddlewareSelection where
+  MiddlewareSelection f <> MiddlewareSelection g =
+    MiddlewareSelection $ \req -> f req <|> g req
+
+instance Monoid MiddlewareSelection where
+  mempty = MiddlewareSelection $ const $ Just passthroughMiddleware
```
Comment (snoyberg) [inline wai-extra/Network/Wai/Middleware/Select.hs:61]: This doesn't look law-abiding to me. It should probably be `const Nothing`. I don't think it's just an issue of laws, I think the current implementation could lead to surprising behavior.

Comment (blackheaven) [reply wai-extra/Network/Wai/Middleware/Select.hs]: Good catch, thanks

File: wai-extra/Network/Wai/Middleware/Select.hs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+-- * Helpers
+--------------------------------------------------
+
+passthroughMiddleware :: Middleware
+passthroughMiddleware = id
+
+-- | Use the 'Middleware' when the predicate holds.
+selectMiddlewareOn :: (Request -> Bool) -> Middleware -> MiddlewareSelection
+selectMiddlewareOn doesApplypath mw = MiddlewareSelection $ \request ->
+  if doesApplypath request
```
Comment (snoyberg) [inline wai-extra/Network/Wai/Middleware/Select.hs:81]: I don't think `path` is needed in `doesApplypath`, this predicate can work for any kind of test.

**Review by snoyberg — CHANGES_REQUESTED** [review]

File: wai-extra/Network/Wai/Middleware/Select.hs (hunk 2 above)
Comment (blackheaven) [reply wai-extra/Network/Wai/Middleware/Select.hs]: Actually, it was a remaining of a refactoring that obviously went wrong :/

**Pull request merged**